README.md
# vscoder <img src="man/figure/vscoder.png" align="right" width="120" />
![lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-lightgrey)![status:
development](https://img.shields.io/badge/status-development-blue)

# Overview

Some functions are provided for the workflow between the R programming language and the VS Code editor.

# Installation

You can install the released version of vscoder from [CRAN](https://CRAN.R-project.org) with:

- CRAN
``` r
# Wait.
```

- Github

Or install the development version of [Github](https://github.com/):
``` r
remotes::install_github("brunofariadf/vscoder")
```

# Usage

This is a basic example that shows how to use a package function.

- R
```r
# version
vscoder::vscode_version()
# [1] "1.69.2"
```

```r
# extension
vscoder::vscode_extension(type = "info")
#                   id           version           license         publisher
# 1  hediet.vscode-dr~              <NA>              <NA>              <NA>
# 2  hediet.vscode-dr~             1.6.4           GPL-3.0 Henning Dieteric~
# 3  julialang.langua~              <NA>              <NA>              <NA>
# 4  julialang.langua~            1.6.28 SEE LICENSE IN L~         julialang
# 5  ms-python.python~         2022.10.1               MIT         Microsoft
# 6  ms-python.vscode~         2022.7.40 SEE LICENSE IN L~         Microsoft
# 7  ms-toolsai.jupyt~ 2022.6.120198181~               MIT         Microsoft
# 8  ms-toolsai.jupyt~             1.0.0               MIT         Microsoft
# 9  ms-toolsai.jupyt~             1.0.9               MIT         Microsoft
# 10 ms-toolsai.jupyt~             1.0.0               MIT         Microsoft
# 11 ms-toolsai.jupyt~             1.0.9               MIT         Microsoft
# 12 ms-vscode.powers~          2022.6.3 SEE LICENSE IN L~         Microsoft
# 13 quarto.quarto-1.~            1.29.0               MIT            Quarto
# 14 reditorsupport.r~             2.5.2 SEE LICENSE IN L~    REditorSupport
#          description
# 1               <NA>
# 2  This unofficial ~
# 3               <NA>
# 4  Julia Language S~
# 5  IntelliSense (Py~
# 6  A performant, fe~
# 7  Jupyter notebook~
# 8  Jupyter keymaps ~
# 9  Renderers for Ju~
# 10 Jupyter keymaps ~
# 11 Renderers for Ju~
# 12 Develop PowerShe~
# 13 Extension for th~
# 14 R Extension for ~
```

# Platform

Development.

<div align="center">

|platform | available |
|:--------|:---------:|
|Windows  |    yes    |
|Linux    |    no     |
|MacOS    |    no     |

</div>

# Code of Conduct

Please note that the project is released with a [Contributor Code of Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html). By contributing to this project, you agree to abide by its terms.

# License

[MIT](https://github.com/brunofariadf/vscoder/blob/main/LICENSE.md)


NEWS.md
# vscoder 0.2.6.9000 (2023-09-07)

## Note from version

### &gt; vscoder (development)

-   development

### &gt; vscoder 0.2.6.9000 (2023-09-07)

- Fix for error message if VS Code editor is not installed.
- Correction for documentation of the functions.

### &gt; vscoder 0.2.4.9000 (2022-07-31)

-   Internal package improvements.
    -   `vscode_extension()` aplication correct from format argument to
        ‘basic’ type argument.
    -   Removed the old error system.

### &gt; vscoder 0.2.2.9000 (2022-07-28)

-   Added function `exists_vscode()` to search installed editor.
-   `vscode_extension()` received two arguments: type, a character class
    vector that indicates ‘basic’ or ‘extended’ information about
    installed extension; and ‘format’, a logical class vector that
    indicates whether the return of the function should have abbreviated
    lines.
-   Internal package improvements.
    -   The internal error system of the package has been redone for the
        purpose of maintaining long-term code.
    -   The package’s internal version control system has been automated
        for the purpose of maintaining long-term code.

### &gt; vscoder 0.0.0.9000 (2022-07-22)

-   Start of the package.


R/main.R
## ------------------------------------------
##                          Date: 2022-07-21
## ------------------------------------------
## Project: vscoder
## Author: Bruno Faria
## 
## GUI: vscode (1.69.2)
## Version R: 4.1.0
## Platform: windows (10.0.14393)
## ------------------------------------------

#' Check the VS Code Editor
#'
#' Check if the VS Code editor exists.
#'
#' @usage exists_vscode()
#' @return The logical class vector indicates if VS Code editor exists.
#' @export 
#' @examples
#' # default
#' # exists_vscode()

exists_vscode <- function() {
    .check_exists_vscode()
}

#' VS Code Version
#'
#' Get the VS Code editor version.
#'
#' @usage vscode_version()
#' @return character class vector that indicates the version of the VS Code editor.
#' @export 
#' @examples
#' # default
#' # vscode_version()

vscode_version <- function() {
    .check_vscode_version()
}

#' VS Code Extension
#'
#' Get the extension that was installed in the VS Code editor.
#'
#' @usage vscode_extension(type = "basic", format = TRUE)
#' @param type character class vector that indicates the type of search to get the extension with options 'basic' and 'info'.
#' @param format logical class vector that indicates whether the column should be formatted with 'TRUE' or 'FALSE' option.
#' @return The object of the dataframe class with information from the VS Code editor extension.
#' @export 
#' @examples
#' # default
#' # vscode_extension()
#' # vscode_extension(type = "info")

vscode_extension <- function(type = "basic", format = TRUE) {
    .check_vscode_extension(type, format)
}

#' VS Code Extension Path
#'
#' Get the VS Code editor extension path.
#'
#' @usage path_extension(extension)
#' @param extension character class vector that indicates the id of the extension.
#' @return dataframe class object with VS Code editor extension path information.
#' @export 
#' @examples
#' path_extension("reditorsupport.r")

path_extension <- function(extension) {
    .check_path_extension(extension)
}


R/check.R
## ------------------------------------------
##                          Date: 2022-07-21
## ------------------------------------------
## Project: vscoder
## Author: Bruno Faria
## 
## GUI: vscode (1.69.2)
## Version R: 4.1.0
## Platform: windows (10.0.14393)
## ------------------------------------------

# check vscode
# not export

.check_exists_vscode <- function() {
    if (!.is_windows()) {
        stop(.error("platform", NULL))
    }

    .get_exists_vscode()
}

# check version
# not export

.check_vscode_version <- function() {
    
    if (!.is_windows()) {
        stop(.error("platform", NULL))
    }

    if (!exists_vscode()) {
        stop(.error("code", NULL))
    }

    .get_version()
}

# check extension
# not export

.check_vscode_extension <- function(type, format) {
    if (!.is_windows()) {
        stop(.error("platform", NULL))
    }

    if (!exists_vscode()) {
        stop(.error("code", NULL))
    }

    if (!is.character(type)) {
        stop(.error("character", type))
    }

    if (!is.element(type, c("basic", "info"))) {
        stop(.error("type_extension", type))
    }

    if (!is.logical(format)) {
        stop(.error("logical", format))
    }

    if (!is.element(format, c("TRUE", "FALSE"))) {
        stop(.error("true_false", format))
    }

    .get_extension(type, format)
}

# check extension path
# not export

.check_path_extension <- function(extension) {
    if (!.is_windows()) {
        stop(.error("platform", NULL))
    }

    if (!exists_vscode()) {
        stop(.error("code", NULL))
    }

    if (!is.character(extension)) {
        stop(.error("character", extension))
    }

    .get_path_extension(extension)
}


R/error.R
## ------------------------------------------
##                          Date: 2022-07-28
## ------------------------------------------
## Project: vscoder
## Author: Bruno Faria
## 
## GUI: VSCode (1.69.2)
## Language: R (4.1.0)
## Platform: Windows
## ------------------------------------------

# check error
# not export

.error_platform <- function(insert) {
    p_i <- .Platform[["OS.type"]]
    paste0("Method not applied to the ", p_i, " platform.")
}

.error_code <- function() {
    paste0("The vs-code installation was not found.")
}

.error_character <- function(insert) {
    paste0("'", insert, "' must be character class vector.")
}

.error_logical <- function(insert) {
    paste0("'", insert, "' must be logical class vector.")
}

.error_numeric <- function(insert) {
    paste0("'", insert, "' must be numeric class vector.")
}

.error_type_extension <- function(insert) {
    paste0("'", insert, "' must be numeric class vector with option 'basic' and 'info'.")
}

.error_true_false <- function(insert) {
    paste0("'", insert, "' must be numeric class vector with option 'TRUE' and 'FALSE'.")
}

.error <- function(type, insert) {
    switch(type,
        "platform" = .error_platform(insert),
        "code" = .error_code(),
        "character" = .error_character(insert),
        "logical" = .error_logical(insert),
        "type_extension" = .error_type_extension(insert),
        "true_false" = .error_true_false(insert),
        "Unknown error type."
    )
}


R/utils.R
## ------------------------------------------
##                          Date: 2022-07-22
## ------------------------------------------
## Project: vscoder
## Author: Bruno Faria
## 
## GUI: vscode (1.69.2)
## Version R: 4.1.0
## Platform: windows (10.0.14393)
## ------------------------------------------

## format with many characters
## not export

.format_character_length <- function(x){
  # select and condition
  i_len <- function(x) {
    x <- unlist(strsplit(x, ""))[1:16]
    x <- paste(c(x, "~"), sep = "", collapse = "")
    return(x)
  }
  
  # limit
  i_pos <- function(x) {
    nchar(x) > 16
  }

  # check
  check_i <- function(x) {
      if (i_pos(x)) i_len(x) else x
  }

  execute_i <- function(x) {
    if (length(x) >= 1 && is.na(x)) {
      return(x)
    } else {
        x_i <- tryCatch(unlist(lapply(x, check_i)), error = function(i)NA)
        return(x_i)
    }
  }

  unlist(lapply(x, execute_i))
}

# format system
# not export

.shell <- function(cmd) {
  obj_i <- suppressWarnings(shell(cmd, 
    intern = TRUE,
    ignore.stderr = TRUE))
  if (length(obj_i) >= 1) {
    return(obj_i)
  } else {
    return(NA)
  }
}

# check version
# not export

.count_version <- function(path, date, dev = TRUE) {
    .version <- function(x) {
        if (unique(x) == 0) {
            return(0)
        } else {
            return(max(x))
        }
    }

    if (file.exists(path)) {
        if (date == "start") {
            return("0.0.0.9000")
        } else {
            df_i <- utils::read.csv2(path, sep = ",")
            df_i <- if (date == "all") df_i else df_i[df_i[["date"]] <= as.Date(date),]
            major_i <- .version(df_i[["version"]])
            minor_i <- sum(grepl(df_i[["type"]], pattern = "minor"))
            patch_i <- sum(grepl(df_i[["type"]], pattern = "patch"))
            dev_i <- if (isTRUE(dev)) ".9000" else ""
            paste0(major_i, ".", minor_i, ".", patch_i, dev_i, " (", date, ")")
        }
    } else {
        stop("unknown file path.")
    }
}


inst/scripts/build.R
## ------------------------------------------
##                          Date: 2022-07-28
## ------------------------------------------
## Project: vscoder
## Author: Bruno Faria
## 
## GUI: VSCode (1.69.2)
## Language: R (4.1.0)
## Platform: Windows
## ------------------------------------------

## --------------------------------------------------------------------------------
## set variable - date: 2022-07-28 15:44:11
## --------------------------------------------------------------------------------

path_csv <- file.path("inst", "exdata", "version", "news.csv")
path_rmd <- file.path("inst", "exdata", "version", "news.Rmd")
path_md <- file.path("inst", "exdata", "version", "NEWS.md")
path_news <- file.path(getwd(), "NEWS.md")

## --------------------------------------------------------------------------------
## News - date: 2022-07-28 15:47:39
## --------------------------------------------------------------------------------

rmarkdown::render(input = path_rmd, output_file = path_news)

## ------------------------------------------
## remove
## ------------------------------------------

# fs::file_move(path_md, getwd())
rm(path_csv, path_md, path_news, path_rmd)


R/engineer.R
## ------------------------------------------
##                          Date: 2022-07-21
## ------------------------------------------
## Project: vscoder
## Author: Bruno Faria
## 
## GUI: vscode (1.69.2)
## Version R: 4.1.0
## Platform: windows (10.0.14393)
## ------------------------------------------

# get vscode
# not export

.get_exists_vscode <- function() {
    if (.is_windows()) {
        x_i <- .shell("where code")
        if (!anyNA(x_i)) {
            x_i <- file.exists(x_i)
            is.element(TRUE, x_i)
        } else {
            return(FALSE)
        }
    }
}

# get version
# not export

.get_version <- function() {
    if (.is_windows()) {
        .shell("code --version")[1]
    }
}

# get extension
# not export

.get_extension <- function(type, format) {
    if (.is_windows()) {
        if (type == "basic") {
            id_i <- .get_id()
            version_i <- .get_id_version()
            id_i <- if (isTRUE(format)) .format_character_length(id_i) else id_i
            version_i <- if (isTRUE(format)) .format_character_length(version_i) else version_i
            data.frame(id = id_i, version = version_i)
        } else if (type == "info") {
            path_i <- .get_path_extension_package(.get_id())
            name_i <- basename(dirname(path_i))
            df_i <- lapply(path_i, 
                function(x)tryCatch(.df_list(jsonlite::fromJSON(x)),
                    error = function(i).df_list_error(x)))
            df_i <- if (isTRUE(format)) lapply(df_i, .format_character_length) else df_i
            df_i <- do.call(rbind, df_i)
            id_i <- if (isTRUE(format)) .format_character_length(name_i) else name_i
            data.frame(id = id_i, df_i)
        }
    }
}

# get extension path
# not export

.get_path_extension <- function(extension) {
    path_i <- .list_extension_path()
    Filter(function(x)grepl(x, pattern = extension), path_i)
}

.get_path_extension_package <- function(extension) {
    pkg_i <- paste0("package", ".", "json")
    path_i <- unlist(lapply(extension, path_extension))
    suppressWarnings(normalizePath(file.path(path_i, pkg_i)))
}

## ------------------------------------------
## engineer utils
## ------------------------------------------

# get id
# not export

.get_id <- function() {
    if (.is_windows()) {
        id_i <- .shell("code --list-extensions")
        tolower(id_i)
    }
}

# get id version
# not export

.get_id_version <- function() {
    if (.is_windows()) {
        ext_ver <- .shell("code --list-extensions --show-versions")
        ext_ver <- strsplit(ext_ver, split = "@")
        as.vector(unlist(lapply(ext_ver, function(x)x[2])))
    }
}

# list extension path
# not export

.list_extension_path <- function() {
    if (.is_windows()) {
        m_i <- Sys.getenv("USERPROFILE")
        m_i <- file.path(m_i, ".vscode", "extensions")
        list_i <- list.dirs(m_i, recursive = FALSE)
        list_i <- as.vector(unlist(
            lapply(list_i, 
                function(x)suppressWarnings(
                    normalizePath(list_i)))
        ))
        unique(list_i)
    }
}

# select df to extension
# not export

.df_list <- function(data) {
    data.frame(
        version = data[["version"]],
        license = data[["license"]],
        publisher = data[["__metadata"]][["publisherDisplayName"]],
        description = data[["description"]]
    )
}

.df_list_error <- function(data) {
    data.frame(
        version = NA,
        license = NA,
        publisher = NA,
        description = NA
    )
}
